extensions/amp-form/0.1/amp-form.js

```diff
@@ -652,6 +736,33 @@ export class AmpForm {
     return this.waitOnPromisesOrTimeout_(varSubPromises, 100);
   }
 
+  /**
+   * Call getValue() on Async Input elements, and
+   * Create hidden inputs containing their returned values
+   * @param {!Element} asyncInput
+   * @return {!Promise}
+   * @private
+   */
+  getValueForAsyncInput_(asyncInput) {
+    return asyncInput.getImpl()
+        .then(implementation => implementation.getValue())
+        .then(value => {
+          const name = asyncInput.getAttribute(AsyncInputAttributes.NAME);
+          let input = this.form_.querySelector(
+              `input[name=${escapeCssSelectorIdent(name)}]`
+          );
+          if (!input) {
+            input = document.createElement('input');
```

Comment:
use `this.win_.document` for easier testing, and prevent cases where `this.win_ != window`. 
And please use `#createElementsWithAttributes`.

Reply:
+1, never use raw `document` or `window` references in AMP components. We should really make this a linter rule.